gomock/call.go

```diff
@@ -75,6 +75,9 @@ func (c *Call) Do(f interface{}) *Call {
 	return c
 }
 
+// Return sets the values to return when the target method is called.
+//
+// If Return and Do are called, the values returned via Do will be returned.
```

Comment:
> A panic for mismatched function signature would be a gift, as it would have exposed the error in my code.

This is actually noted in the current code as a standing TODO, annoyingly: https://github.com/golang/mock/blob/master/gomock/call.go#L73

Definitely something that needs to actually be implemented, though, as Do _already_ has a standing panic case if the DoFunc doesn't have the same input argument signature as the parent function.  I just didn't really want to see _another_ panic case added to upstream, particularly when that panic case is triggered by the majority current usage of Do (ie. with functions that don't return anything).

> Perhaps, we should define a new method, Stub for this functionality.

That's actually already been suggested and rejected, unfortunately: https://github.com/golang/mock/issues/24

Though your usage of the term seems to suggest that you had a different mechanic in mind, since that type of "Stub" wouldn't fix the issue this PR is attempting to address.

Personally, I think the mechanic you're going for, the ability to process the inputs to a call and alter the return values based on that (what I believe you're referring to when you say "stub"), is something beyond the scope of what `Do()` is intended for, and currently used for.  As such, it might be best to make a new method entirely (say, `Process()`), rather than attempt to piggyback the behavior onto the existing `Do()` method.  This would also be a good place to do the arity and type checking for both input and output arguments, which could by copied over to `Do()` as well.  I'll see if I can't cook up some code to do that checking.

> I considered this. Either behavior is problematic, as it may result in test failures for existing uses, making the change backwards compatible.

I'd actually argue that `Return()` overriding the return from `Do()` matches current behavior better.  Right now, if both `Do()` and `Return()` are called, the values passed to `Return()` are what are actually returned, because `Do()` has no return behavior (it's return values aren't even stored).  As such, changing it so `Do()` overrides `Return()` (as exists in this PR) will break those existing statement, and will also trigger the panic issue on those calls.  However, if `Return()` overrides `Do()`, those cases don't break (and no other cases break instead, either), and are in fact protected from the return type/arity mismatch panic as well.

Also worth noting that the current functionality of "on call, do this thing, and also return these other things" is impossible with the current implementation in this PR unless your "do this thing" function returns those "other things", making test cases more complicated than they need to be.  Making `Return()` override `Do()`, however, removes no existing functionality.

Here are the existing possible combinations for the parent function `Foo(a A) B`:
* `EXPECT().Foo(a).Returns(b)` - no change from this PR either way
* `EXPECT().Foo(a).Do(func(a A) B { return b })` - changed so the `B` returned by the lambda passed to `Do()` is actually processed and then returned by the `Foo()` call itself.  This occurs regardless of the `Do()`/`Return()` resolution.
* `EXPECT().Foo(a).Returns(b).Do(func(a A) B { return b2 })` - this is the problem case.  Currently, this returns `b`.  If we use this PR's current setup, the call to `Return()` is completely ignored, and `b2` is returned, despite the programmer clearly intending `b` to be returned.  If we make `Return()` override `Do()`, the existing semantic of `b` being returned persists, with the new behavior of returning the DoFunc return value only occurring if one does not _explicitly_ indicate the desired return value.

Though, again, I think this would be a better implementation if it weren't piggybacked on top of `Do()`, and instead was a separate (and new) method call like `Process()`.